app/Http/Controllers/Auth/LoginController.php

```diff
@@ -36,161 +28,78 @@ class LoginController extends BaseController
         }
-        throw new \Exception('SSO failed: '.VatsimSSO::error()['message']);
-    }
-
-    protected function setVatsimAuth($userId)
-    {
-        Auth::guard('vatsim-sso')->loginUsingId($userId);
-    }
 
-    protected function attemptSecondaryAuth()
-    {
-        $member = Auth::guard('vatsim-sso')->user();
-
-        if ($member->hasPassword()) {
-            return redirect()->route('auth-secondary');
-        }
+        $account = $this->completeLogin($resourceOwner, $accessToken);
 
-        $intended = Session::pull('url.intended', route('site.home'));
+        Auth::guard('vatsim-sso')->loginUsingId($account->id);
 
-        Auth::login(Auth::guard('vatsim-sso')->user(), true);
-
-        return redirect($intended);
+        return SecondaryLoginController::attemptSecondaryAuth($account);
     }
 
-    public function loginSecondary(Request $request)
+    protected function completeLogin($resourceOwner, $token)
     {
-        if (! Auth::guard('vatsim-sso')->check()) {
-            return redirect()->route('dashboard')
-                ->withError('Could not authenticate: VATSIM.net authentication is not present.');
-        }
-
-        Auth::shouldUse('web');
-        $response = $this->login($request);
-
-        return $response;
-    }
-
-    /**
-     * Get the needed authorization credentials from the request.
-     *
-     * @param  \Illuminate\Http\Request $request
-     * @return array
-     */
-    protected function credentials(Request $request)
-    {
-        return ['id' => Auth::guard('vatsim-sso')->user()->id, 'password' => $request->input('password')];
-    }
-
-    /**
-     * Validate the user login request.
-     *
-     * @param  \Illuminate\Http\Request $request
-     * @return void
-     */
-    protected function validateLogin(Request $request)
-    {
-        $this->validate($request, [
-            'password' => 'required|string',
-        ]);
-    }
-
-    public function vatsimSsoReturn(Request $request)
-    {
-        $ssoCredentials = $request->session()->remove('credentials.vatsim-sso');
-
-        $login = VatsimSSO::checkLogin($ssoCredentials['key'], $ssoCredentials['secret'], $request->input('oauth_verifier'));
-        if ($login) {
-            return $this->vSsoValidationSuccess($login->user, $login->request);
-        } else {
-            return $this->vSsoValidationFailure(VatsimSSO::error());
-        }
-    }
-
-    public function vSsoValidationSuccess($user, $request)
-    {
-        $account = Account::firstOrNew(['id' => $user->id]);
-        $account->name_first = utf8_decode($user->name_first);
-        $account->name_last = utf8_decode($user->name_last);
-        $account->email = $user->email;
-        $account->experience = $user->experience;
-        $account->joined_at = $user->reg_date;
+        $account = Account::firstOrNew(['id' => $resourceOwner->data->cid]);
```

Comment:
```php
[$name_first, $name_last, $email] = $resourceOwner->data;
```

Reply:
Tried playing around with this and seemed to cause a mild faff. Will stick with it as is.